profiling/src/io.rs

```diff
@@ -0,0 +1,683 @@
+        IOProfilingStats::new(
+            SOCKET_READ_TIME_PROFILING_INTERVAL.load(Ordering::SeqCst) as f64,
+            SocketReadTimeCollector
+        )
+    );
+    static SOCKET_WRITE_TIME_PROFILING_STATS: RefCell<IOProfilingStats<SocketWriteTimeCollector>> = RefCell::new(
+        IOProfilingStats::new(
+            SOCKET_WRITE_TIME_PROFILING_INTERVAL.load(Ordering::SeqCst) as f64,
+            SocketWriteTimeCollector
+        )
+    );
+    static FILE_READ_TIME_PROFILING_STATS: RefCell<IOProfilingStats<FileReadTimeCollector>> = RefCell::new(
+        IOProfilingStats::new(
+            FILE_READ_TIME_PROFILING_INTERVAL.load(Ordering::SeqCst) as f64,
+            FileReadTimeCollector
+        )
+    );
+    static FILE_WRITE_TIME_PROFILING_STATS: RefCell<IOProfilingStats<FileWriteTimeCollector>> = RefCell::new(
+        IOProfilingStats::new(
+            FILE_WRITE_TIME_PROFILING_INTERVAL.load(Ordering::SeqCst) as f64,
+            FileWriteTimeCollector
+        )
+    );
+    static SOCKET_READ_SIZE_PROFILING_STATS: RefCell<IOProfilingStats<SocketReadSizeCollector>> = RefCell::new(
+        IOProfilingStats::new(
+            SOCKET_READ_SIZE_PROFILING_INTERVAL.load(Ordering::SeqCst) as f64,
+            SocketReadSizeCollector
+        )
+    );
+    static SOCKET_WRITE_SIZE_PROFILING_STATS: RefCell<IOProfilingStats<SocketWriteSizeCollector>> = RefCell::new(
+        IOProfilingStats::new(
+            SOCKET_WRITE_SIZE_PROFILING_INTERVAL.load(Ordering::SeqCst) as f64,
+            SocketWriteSizeCollector
+        )
+    );
+    static FILE_READ_SIZE_PROFILING_STATS: RefCell<IOProfilingStats<FileReadSizeCollector>> = RefCell::new(
+        IOProfilingStats::new(
+            FILE_READ_SIZE_PROFILING_INTERVAL.load(Ordering::SeqCst) as f64,
+            FileReadSizeCollector
+        )
+    );
+    static FILE_WRITE_SIZE_PROFILING_STATS: RefCell<IOProfilingStats<FileWriteSizeCollector>> = RefCell::new(
+        IOProfilingStats::new(
+            FILE_WRITE_SIZE_PROFILING_INTERVAL.load(Ordering::SeqCst) as f64,
+            FileWriteSizeCollector
+        )
+    );
+}
+
+pub fn io_prof_first_rinit() {
+    let io_profiling = REQUEST_LOCALS.with(|cell| {
+        cell.try_borrow()
+            .map(|locals| locals.system_settings().profiling_io_enabled)
+            .unwrap_or(false)
+    });
+
+    trace!("I/O profiling: {io_profiling:?}");
+
+    if io_profiling {
+        unsafe {
+            let mut overwrites = vec![
+                GotSymbolOverwrite {
+                    symbol_name: "recv",
+                    new_func: observed_recv as *mut (),
+                    orig_func: ptr::addr_of_mut!(ORIG_RECV) as *mut _ as *mut *mut (),
+                },
+                GotSymbolOverwrite {
+                    symbol_name: "send",
+                    new_func: observed_send as *mut (),
+                    orig_func: ptr::addr_of_mut!(ORIG_SEND) as *mut _ as *mut *mut (),
+                },
```

Comment:
Is there a reason we don't monitor the sendmsg and recvmsg from ext/sockets?

Reply:
done in 01eb7c1